java/tools/src/main/java/org/apache/arrow/tools/FileToStream.java

```diff
@@ -38,11 +38,21 @@
   public static void convert(FileInputStream in, OutputStream out) throws IOException {
     BufferAllocator allocator = new RootAllocator(Integer.MAX_VALUE);
     try (ArrowFileReader reader = new ArrowFileReader(in.getChannel(), allocator)) {
-      try (ArrowStreamWriter writer = new ArrowStreamWriter(reader.getVectorSchemaRoot(), reader, out)) {
-        for (ArrowBlock block: reader.getRecordBlocks()) {
-          reader.loadRecordBatch(block);
-          writer.writeBatch();
+      VectorSchemaRoot root = reader.getVectorSchemaRoot();
+      // load the first batch before instantiating the writer so that we have any dictionaries
+      reader.loadNextBatch();
+      try (ArrowStreamWriter writer = new ArrowStreamWriter(root, reader, out)) {
+        writer.start();
+        while (true) {
+          int rowCount = reader.getVectorSchemaRoot().getRowCount();
+          if (rowCount == 0) {
+            break;
+          } else {
+            writer.writeBatch();
+            reader.loadNextBatch();
```

Comment:
If you had a array or list of in-memory record batches in Java, can you write them with this API? I was able to add dictionaries to the C++ schemas without modifying the stream / file API, and schemas are immutable. I think it would be worth working toward that eventually, maybe we should open some JIRAs to cover

* Making Java schemas immutable
* Looser coupling between the reader and writer classes here